### PR TITLE
Initialize class variables with default values in scene/ [1/2]

### PR DESCRIPTION
**Pull request by qarmin**
Part of #43636
Fixes in 4.0 https://github.com/godotengine/godot/issues/35877

Also moved initialization of data from constructors.

### CODE REVIEW

**Review by pouleyKetchoupp — APPROVED** [review]
Physics classes look fine (I haven't checked other classes).

File: scene/2d/animated_sprite_2d.h [hunk 1]
```diff
@@ -38,14 +38,9 @@ class SpriteFrames : public Resource {
 	GDCLASS(SpriteFrames, Resource);
 
 	struct Anim {
-		float speed;
-		bool loop;
+		float speed = 5;
```
Comment (akien-mga) [inline scene/2d/animated_sprite_2d.h:41]:
```suggestion
		float speed = 5.0;
```
Just to make the `float` explicit while we're at it.

File: scene/2d/animated_sprite_2d.h [hunk 2]
```diff
@@ -109,20 +104,20 @@ class AnimatedSprite2D : public Node2D {
...
+	float speed_scale = 1.0f;
 
-	bool centered;
+	bool centered = true;
 	Point2 offset;
 
-	bool is_over;
-	float timeout;
+	bool is_over = false;
+	float timeout = 0;
```
Comment (akien-mga) [inline scene/2d/animated_sprite_2d.h:117]:
```suggestion
	float timeout = 0.0;
```

File: scene/2d/area_2d.h [hunk 3]
```diff
@@ -47,28 +47,28 @@ class Area2D : public CollisionObject2D {
...
-	uint32_t collision_mask;
-	uint32_t collision_layer;
-	int priority;
-	bool monitoring;
-	bool monitorable;
-	bool locked;
+	bool gravity_is_point = false;
+	real_t gravity_distance_scale = 0;
+	real_t linear_damp = 0.1;
+	real_t angular_damp = 1;
```
Comment (akien-mga) [inline scene/2d/area_2d.h:56]:
```suggestion
	real_t gravity_distance_scale = 0.0;
	real_t linear_damp = 0.1;
	real_t angular_damp = 1.0;
```

File: scene/2d/audio_stream_player_2d.h [hunk 4]
```diff
@@ -47,32 +47,32 @@ class AudioStreamPlayer2D : public Node2D {
...
 	Ref<AudioStreamPlayback> stream_playback;
 	Ref<AudioStream> stream;
 	Vector<AudioFrame> mix_buffer;
 
-	volatile float setseek;
-	volatile bool active;
-	volatile float setplay;
+	volatile float setseek = -1;
+	volatile bool active = false;
+	volatile float setplay = -1;
```
Comment (akien-mga) [inline scene/2d/audio_stream_player_2d.h:68]:
```suggestion
	volatile float setseek = -1.0;
	volatile bool active = false;
	volatile float setplay = -1.0;
```

File: scene/2d/audio_stream_player_2d.h [hunk 5]
```diff
@@ -47,32 +47,32 @@ class AudioStreamPlayer2D : public Node2D {
...
 
-	float volume_db;
-	float pitch_scale;
-	bool autoplay;
-	bool stream_paused;
-	bool stream_paused_fade_in;
-	bool stream_paused_fade_out;
+	float volume_db = 0;
+	float pitch_scale = 1;
```
Comment (akien-mga) [inline scene/2d/audio_stream_player_2d.h:71]:
```suggestion
	float volume_db = 0.0;
	float pitch_scale = 1.0;
```

File: scene/2d/camera_2d.h [hunk 6]
```diff
@@ -49,24 +49,24 @@ class Camera2D : public Node2D {
 	};
 
 protected:
-	Point2 camera_pos;
+	Point2 camera_pos = Vector2();
```
Comment (akien-mga) [inline scene/2d/camera_2d.h:52]: `Point2` is already default initialized IINM.

File: scene/2d/camera_2d.cpp [hunk 7]
```diff
@@ -749,9 +749,6 @@ void Camera2D::_bind_methods() {
 }
 
 Camera2D::Camera2D() {
```
Comment (akien-mga) [inline scene/2d/camera_2d.cpp:751]: You forgot to remove some assignments in this constructor.

File: scene/2d/collision_object_2d.h [hunk 8]
```diff
@@ -37,35 +37,28 @@
...
-
-		ShapeData() {
-			disabled = false;
-			one_way_collision = false;
-			one_way_collision_margin = 0;
-			owner = nullptr;
-		}
+		bool disabled = false;
+		bool one_way_collision = false;
+		float one_way_collision_margin = 0;
```
Comment (akien-mga) [inline scene/2d/collision_object_2d.h:55]:
```suggestion
		float one_way_collision_margin = 0.0;
```

File: scene/2d/collision_polygon_2d.h [hunk 9]
```diff
@@ -46,14 +46,14 @@ class CollisionPolygon2D : public Node2D {
...
-	uint32_t owner_id;
-	CollisionObject2D *parent;
-	bool disabled;
-	bool one_way_collision;
-	float one_way_collision_margin;
+	uint32_t owner_id = 0;
+	CollisionObject2D *parent = nullptr;
+	bool disabled = false;
+	bool one_way_collision = false;
+	float one_way_collision_margin = 1;
```
Comment (akien-mga) [inline scene/2d/collision_polygon_2d.h:56]:
```suggestion
	float one_way_collision_margin = 1.0;
```

File: scene/2d/collision_shape_2d.h [hunk 10]
```diff
@@ -39,13 +39,13 @@ class CollisionObject2D;
...
+	Rect2 rect = Rect2(-Point2(10, 10), Point2(20, 20));
+	uint32_t owner_id = 0;
+	CollisionObject2D *parent = nullptr;
 	void _shape_changed();
-	bool disabled;
-	bool one_way_collision;
-	float one_way_collision_margin;
+	bool disabled = false;
+	bool one_way_collision = false;
+	float one_way_collision_margin = 1;
```
Comment (akien-mga) [inline scene/2d/collision_shape_2d.h:48]:
```suggestion
	float one_way_collision_margin = 1.0;
```

File: scene/2d/cpu_particles_2d.h [hunk 11]
```diff
@@ -83,19 +83,19 @@ class CPUParticles2D : public Node2D {
 	struct Particle {
 		Transform2D transform;
 		Color color;
-		float custom[4];
-		float rotation;
+		float custom[4] = {};
+		float rotation = 0;
```
Comment (akien-mga) [inline scene/2d/cpu_particles_2d.h:87]:
```suggestion
		float rotation = 0.0;
```

File: scene/2d/cpu_particles_2d.h [hunk 12]
```diff
@@ -83,19 +83,19 @@ class CPUParticles2D : public Node2D {
...
-		float anim_offset_rand;
-		float time;
-		float lifetime;
+		bool active = false;
+		float angle_rand = 0;
+		float scale_rand = 0;
+		float hue_rot_rand = 0;
+		float anim_offset_rand = 0;
+		float time = 0;
+		float lifetime = 0;
```
Comment (akien-mga) [inline scene/2d/cpu_particles_2d.h:95]:
```suggestion
		float angle_rand = 0.0;
		float scale_rand = 0.0;
		float hue_rot_rand = 0.0;
		float anim_offset_rand = 0.0;
		float time = 0.0;
		float lifetime = 0.0;
```

File: scene/2d/cpu_particles_2d.h [hunk 13]
```diff
@@ -83,19 +83,19 @@ class CPUParticles2D : public Node2D {
...
 		Color base_color;
 
-		uint32_t seed;
+		uint32_t seed = 0;
 	};
 
 	float time;
```
Comment (akien-mga) [inline scene/2d/cpu_particles_2d.h:101]: You can update some of those too with the values from the constructor.

File: scene/2d/joints_2d.h [hunk 14]
```diff
@@ -43,9 +43,9 @@ class Joint2D : public Node2D {
 
 	NodePath a;
 	NodePath b;
-	real_t bias;
+	real_t bias = 0;
```
Comment (akien-mga) [inline scene/2d/joints_2d.h:46]:
```suggestion
	real_t bias = 0.0;
```

File: scene/2d/joints_2d.h [hunk 15]
```diff
@@ -80,7 +80,7 @@ class Joint2D : public Node2D {
 class PinJoint2D : public Joint2D {
 	GDCLASS(PinJoint2D, Joint2D);
 
-	real_t softness;
+	real_t softness = 0;
```
Comment (akien-mga) [inline scene/2d/joints_2d.h:83]:
```suggestion
	real_t softness = 0.0;
```

File: scene/2d/joints_2d.h [hunk 16]
```diff
@@ -97,8 +97,8 @@ class PinJoint2D : public Joint2D {
 class GrooveJoint2D : public Joint2D {
 	GDCLASS(GrooveJoint2D, Joint2D);
 
-	real_t length;
-	real_t initial_offset;
+	real_t length = 50;
+	real_t initial_offset = 25;
```
Comment (akien-mga) [inline scene/2d/joints_2d.h:101]:
```suggestion
	real_t length = 50.0;
	real_t initial_offset = 25.0;
```

File: scene/2d/joints_2d.h [hunk 17]
```diff
@@ -118,10 +118,10 @@ class GrooveJoint2D : public Joint2D {
...
 	GDCLASS(DampedSpringJoint2D, Joint2D);
 
-	real_t stiffness;
-	real_t damping;
-	real_t rest_length;
-	real_t length;
+	real_t stiffness = 20;
+	real_t damping = 1;
+	real_t rest_length = 0;
+	real_t length = 50;
```
Comment (akien-mga) [inline scene/2d/joints_2d.h:124]:
```suggestion
	real_t stiffness = 20.0;
	real_t damping = 1.0;
	real_t rest_length = 0.0;
	real_t length = 50.0;
```

File: scene/2d/parallax_layer.h [hunk 18]
```diff
@@ -38,7 +38,8 @@ class ParallaxLayer : public Node2D {
 
 	Point2 orig_offset;
 	Point2 orig_scale;
-	Size2 motion_scale;
+	Size2 motion_scale = Size2(1, 1);
+	;
```
Comment (akien-mga) [inline scene/2d/parallax_layer.h:42]: Stray semicolon.

File: scene/2d/parallax_layer.h [hunk 19]
```diff
@@ -38,7 +38,8 @@ class ParallaxLayer : public Node2D {
 
 	Point2 orig_offset;
 	Point2 orig_scale;
-	Size2 motion_scale;
+	Size2 motion_scale = Size2(1, 1);
```
Comment (akien-mga) [inline scene/2d/parallax_layer.h:41]: This can be removed from the constructor.

File: scene/3d/camera_3d.h [hunk 20]
```diff
@@ -57,26 +57,27 @@ class Camera3D : public Node3D {
...
-	float fov;
-	float size;
-	Vector2 frustum_offset;
-	float near, far;
-	float v_offset;
-	float h_offset;
-	KeepAspect keep_aspect;
+	float fov = 0;
+	float size = 1;
+	Vector2 frustum_offset = Vector2();
```
Comment (akien-mga) [inline scene/3d/camera_3d.h:68]: Not necessary, it's default initialized.

File: scene/3d/camera_3d.cpp [hunk 21]
```diff
@@ -653,20 +653,7 @@ Vector3 Camera3D::get_doppler_tracked_velocity() const {
...
-	viewport = nullptr;
-	force_change = false;
-	mode = PROJECTION_PERSPECTIVE;
 	set_perspective(75.0, 0.05, 4000.0);
-	keep_aspect = KEEP_HEIGHT;
-	layers = 0xfffff;
-	v_offset = 0;
-	h_offset = 0;
 	RenderingServer::get_singleton()->camera_set_cull_mask(camera, layers);
 	//active=false;
```
Comment (akien-mga) [inline scene/3d/camera_3d.cpp:658]: Could clean this one up while at it.

Also `doppler_tracking` was missed below.

File: scene/3d/skeleton_3d.h [hunk 22]
```diff
@@ -74,57 +74,44 @@ class Skeleton3D : public Node3D {
 	struct Bone {
 		String name;
 
-		bool enabled;
-		int parent;
-		int sort_index; //used for re-sorting process order
+		bool enabled = false;
```
Comment (akien-mga) [inline scene/3d/skeleton_3d.h:77]: Wrong, it defaults to true in the constructor.

File: scene/3d/vehicle_body_3d.h [hunk 23]
```diff
@@ -42,51 +42,51 @@ class VehicleWheel3D : public Node3D {
...
-	real_t m_rollInfluence;
-	real_t m_engineForce;
-	real_t m_brake;
+	real_t m_steering = real_t(0.);
+	real_t m_rotation = real_t(0.);
+	real_t m_deltaRotation = real_t(0.);
+	real_t m_rpm = 0;
+	real_t m_rollInfluence = real_t(0.);
+	real_t m_engineForce = real_t(0.);
+	real_t m_brake = real_t(0.);
```
Comment (akien-mga) [inline scene/3d/vehicle_body_3d.h:73]: I know it was in the constructor but I don't see why the explicit `real_t()` cast is needed.

File: scene/3d/vehicle_body_3d.h [hunk 24]
```diff
@@ -42,51 +42,51 @@ class VehicleWheel3D : public Node3D {
...
-	real_t m_deltaRotation;
-	real_t m_rpm;
-	real_t m_rollInfluence;
-	real_t m_engineForce;
-	real_t m_brake;
+	real_t m_steering = real_t(0.);
+	real_t m_rotation = real_t(0.);
+	real_t m_deltaRotation = real_t(0.);
+	real_t m_rpm = 0;
+	real_t m_rollInfluence = real_t(0.);
```
Comment (akien-mga) [inline scene/3d/vehicle_body_3d.h:71]: This was `0.1` in the constructor.

File: scene/3d/vehicle_body_3d.cpp [hunk 25]
```diff
@@ -948,13 +924,6 @@ void VehicleBody3D::_bind_methods() {
...
 VehicleBody3D::VehicleBody3D() {
-	m_pitchControl = 0;
-	m_currentVehicleSpeedKmHour = real_t(0.);
-	m_steeringValue = real_t(0.);
-
-	engine_force = 0;
-	brake = 0;
-
 	state = nullptr;
 	ccd = false;
```
Comment (akien-mga) [inline scene/3d/vehicle_body_3d.cpp:928]: Could be removed, assuming that they're the same default values as in PhysicsBody3D.

File: scene/animation/animation_blend_tree.h [hunk 26]
```diff
@@ -180,8 +180,9 @@ class AnimationNodeAdd3 : public AnimationNode {
 class AnimationNodeBlend2 : public AnimationNode {
 	GDCLASS(AnimationNodeBlend2, AnimationNode);
 
-	StringName blend_amount;
-	bool sync;
+	StringName blend_amount = "blend_amount";
+	;
```
Comment (akien-mga) [inline scene/animation/animation_blend_tree.h:184]: Stay semicolon.

File: scene/animation/root_motion_view.h [hunk 27]
```diff
@@ -39,12 +39,12 @@ class RootMotionView : public VisualInstance3D {
...
 	NodePath path;
-	float cell_size;
-	float radius;
-	bool use_in_game;
-	Color color;
-	bool first;
-	bool zero_y;
+	float cell_size = 1;
+	float radius = 10;
+	bool use_in_game = false;
```
Comment (akien-mga) [inline scene/animation/root_motion_view.h:44]: Unused, can be removed.

**Review by akien-mga — CHANGES_REQUESTED** [review]
Looks great overall, thanks for the huge amount of work (I did the same for `core` and I know how tedious it is :D).

Found a couple bugs introduced by unwillingly changing the default value of some properties.

I initially commented about floats that should have `.0` to make their type clear, but then stopped at some point as there's a ton of them. Would still be worth changing IMO, possibly by grepping the `git show` for `(float|real_t) .* = `.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! Amazing work!

**Pull request merged**